# ã‚·ãƒŠãƒªã‚ªå†ç”Ÿã‚¨ãƒ³ã‚¸ãƒ³
# TODO ã“ã“ã«ã‚·ãƒŠãƒªã‚ªã‚’èª­å–ã‚‹å‡¦ç†ã‚’ç§»å‹•ã—ãŸã„
extends Node


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒ¡ãƒ¢ãƒªé–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

# å…ˆç¥–ã®è¾æ›¸ã‚­ãƒ£ãƒƒã‚·ãƒ¥
var ancestors = {}

# å…¨å‘½ä»¤ï¼ˆã‚­ãƒ¼ï¼š"å‘½ä»¤å:"ã€€å€¤ï¼šãƒãƒ¼ãƒ‰åï¼‰
var directory_for_instruction_code_and_node_name = null


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# å¤–ãƒ‘ã‚¹é–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# ãƒ—ãƒ­ã‚°ãƒ©ãƒ ã‚ºãƒ»ãƒãƒ–å–å¾—
func hub():
	return MonkeyHelper.find_ancestor_child(
			self,
			"ğŸŒProgrammer/ğŸ›©ï¸Hub",
			self.ancestors)


# å…¨ã¦ã®å‘½ä»¤ã‚³ãƒ¼ãƒ‰ä¸€è¦§
func get_all_instruction_codes():
	if self.directory_for_instruction_code_and_node_name == null:
		self.directory_for_instruction_code_and_node_name = {}	# ã‚­ãƒ¼ï¼šStringName, å€¤ï¼šNone

		MonkeyHelper.search_node_name_begins_with(
				# å‘½ä»¤ã®ãƒãƒ¼ãƒ‰åã¯ `ğŸ“—` ã§å§‹ã¾ã‚‹ã‚‚ã®ã¨ã™ã‚‹
				&"ğŸ“—",
				# æ¢ã™å ´æ‰€
				# æœ¬å½“ã¯ `ğŸŒProgrammer` ãƒãƒ¼ãƒ‰ã®ä¸‹ã®ã©ã“ã‹ã«ã‚ã‚‹ `ğŸ“‚ScenarioPlayer_ğŸ‰VisualNovel` ãƒãƒ¼ãƒ‰ã®ã•ã‚‰ã«ä¸‹ã® `ğŸ“‚Instructions` ãƒãƒ¼ãƒ‰ã®ä¸‹ã‚’æ¢ã—ã¦æ¬²ã—ã„ãŒã€‚
				self.hub().get_programmer(),
				func(child_node):
					# ã‚³ãƒ¼ãƒ‰ã«ãƒãƒ¼ãƒ‰åã‚’ç´ã¥ã‘ã‚‹
					self.directory_for_instruction_code_and_node_name[child_node.code] = child_node.name)

	return self.directory_for_instruction_code_and_node_name


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ä»¥ä¸‹ã€ä¸»è¦ãƒ—ãƒ­ã‚°ãƒ©ãƒ 
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# Called when the node enters the scene tree for the first time.
func _ready():
	pass # Replace with function body.


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta):
	pass


# ã€ŒÂ§ã€ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã®å†ç”Ÿ
func play_section():
	var department_value = self.hub().get_current_department_value()
	var message_window_gui = self.hub().get_current_message_window_gui()

	# å…¨éƒ¨æ¶ˆåŒ–æ¸ˆã¿ã®å ´åˆ
	if self.hub().get_current_section_size_of_scenario() <= message_window_gui.section_item_index:
		print("ï¼»ã‚·ãƒŠãƒªã‚ªå†ç”Ÿã‚¨ãƒ³ã‚¸ãƒ³ï¼½ï¼ˆ" + department_value.name + "ã€€" + department_value.section_name + "ï¼‰ã€€ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã‚’èª­ã¿çµ‚ã‚ã£ã¦ã„ã‚‹")

		# ã‹ã¤ã€ã‚³ãƒ³ãƒ—ãƒªãƒ¼ãƒˆä¸­ã®å ´åˆã€ãƒ¦ãƒ¼ã‚¶ãƒ¼å…¥åŠ›ã‚’å¾…ã¤
		if message_window_gui.statemachine_of_message_window.is_completed():
			print("ï¼»ã‚·ãƒŠãƒªã‚ªå†ç”Ÿã‚¨ãƒ³ã‚¸ãƒ³ï¼½ï¼ˆ" + department_value.name + "ã€€"+ department_value.section_name + "ï¼‰ã€€å…¨æ¶ˆåŒ–æ¸ˆã¿ã ãŒã€ã‚³ãƒ³ãƒ—ãƒªãƒ¼ãƒˆä¸­ã ã‹ã‚‰ã€å‹æ‰‹ã«ä½•ã‚‚ã—ãªã„ã€‚ãƒ¦ãƒ¼ã‚¶ãƒ¼å…¥åŠ›ã‚’å¾…ã¤")
			# è‡ªå‹•ã§ä½•ã‹ã—ãªã„
			return

	# ãƒ‘ãƒ¼ã‚¹ã‚’é–‹å§‹ã—ã¦ã‚ˆããªã„ã‚±ãƒ¼ã‚¹ã‚‚ã‚ã‚‹ãŒï¼Ÿ
	# ãƒãƒƒãƒ•ã‚¡ãƒ¼ãŒæ®‹ã£ã¦ã‚‹ã¨ãã¨ã‹
	if not message_window_gui.has_text_block_buffer():
		# Completed æ™‚ã‚‚ãƒ‘ãƒ¼ã‚¹å§‹ã‚ãŸã‚‰ã‚ˆããªã„
		if not message_window_gui.statemachine_of_message_window.is_completed():
			# TODO é¸æŠè‚¢ã®ã¨ãã‚‚ã‚„ã‚„ã“ã—ã„ãŒ
			print("ï¼»ã‚·ãƒŠãƒªã‚ªå†ç”Ÿã‚¨ãƒ³ã‚¸ãƒ³ï¼½ï¼ˆ" + department_value.name + "ã€€"+ department_value.section_name + "ï¼‰ã€€ãƒ‘ãƒ¼ã‚¹ã‚’é–‹å§‹ã—ã¦ã‚ˆã„ï¼ˆæœ¬å½“ã‹ï¼Ÿï¼‰")
			# ãƒ‘ãƒ¼ã‚¹ã‚’é–‹å§‹ã—ã¦ã‚ˆã„
			department_value.set_parse_lock(false)


# ãƒ‘ãƒ©ã‚°ãƒ©ãƒ•ï¼ˆã‚»ã‚¯ã‚·ãƒ§ãƒ³ã®ã‚¢ã‚¤ãƒ†ãƒ ï¼‰ãŒï¼»ãƒˆæ›¸ãï¼½ã‹ã€ï¼»å°è©ï¼½ã‹ã€ã«ã‚ˆã£ã¦å‡¦ç†ã‚’åˆ†ã‘ã¾ã™
func parse_paragraph(paragraph_text):
		
	# ãƒˆæ›¸ããªã‚‰å®Ÿè¡Œ
	if self.execute_stage_directions(paragraph_text):
		# ã™ã‚Œã°æŠœã‘ã‚‹
		return

	# é¸æŠè‚¢ãªã‚‰è¡¨ç¤º
	if self.print_choices(paragraph_text):
		# ã™ã‚Œã°æŠœã‘ã‚‹
		return

	# é€šå¸¸æ–‡æ›¸ã®è¡¨ç¤º
	self.print_normal_text(paragraph_text)


# é€šå¸¸æ–‡æ›¸ã®è¡¨ç¤º	
func print_normal_text(paragraph_text):
	#print("ï¼»ã‚·ãƒŠãƒªã‚ªã‚¨ãƒ³ã‚¸ãƒ³ï¼½ã€€æº–å‚™ä¸­ã€€é€šå¸¸æ–‡æ›¸ã®è¡¨ç¤º")
	self.hub().get_instruction(&"ğŸ“˜NormalText").do_it(paragraph_text)


# é¸æŠè‚¢ãªã‚‰è¡¨ç¤º
func print_choices(paragraph_text):
	#print("ï¼»ã‚·ãƒŠãƒªã‚ªã‚¨ãƒ³ã‚¸ãƒ³ï¼½ã€€æº–å‚™ä¸­ã€€é¸æŠè‚¢ãªã‚‰è¡¨ç¤º")
	var message_window_gui = self.hub().get_current_message_window_gui()

	# é¸æŠè‚¢ã 
	if message_window_gui.choices_row_numbers != null:
		self.hub().get_instruction(&"ğŸ“˜NormalTextChoice").do_it(paragraph_text)
		return true

	return false


# ãƒˆæ›¸ããªã‚‰å®Ÿè¡Œ
func execute_stage_directions(paragraph_text):
	print("ï¼»ã‚·ãƒŠãƒªã‚ªã‚¨ãƒ³ã‚¸ãƒ³ï¼½ã€€æº–å‚™ä¸­ã€€ãƒˆæ›¸ããªã‚‰å®Ÿè¡Œ")
	
	# ï¼»ãƒˆæ›¸ãï¼½ã‹ã©ã†ã‹åˆ¤å®š
	var first_head_tail = self.hub().split_head_line_or_tail(paragraph_text)
	var first_head = first_head_tail[0].strip_edges()
	var first_tail = first_head_tail[1] 
		
	# ï¼»ãƒˆæ›¸ãï¼½
	# `.strip_edges()` - å…ˆé ­è¡Œã®æœ€åˆã¨ã€æœ€çµ‚è¡Œã®æœ€å¾Œã®è¡¨ç¤ºã•ã‚Œãªã„æ–‡å­—ã‚’æ¶ˆå»
	if first_head.strip_edges() == "!":
		print("ï¼»åŠ©ç›£ï¼½ã€€å‘½ä»¤ãƒ†ã‚­ã‚¹ãƒˆã ï¼š[" + first_tail + "]")

		# ã•ã‚‰ã«å…ˆé ­è¡Œã‚’å–å¾—
		var second_head_tail = self.hub().split_head_line_or_tail(first_tail)
		
		while second_head_tail != null:
			var second_head = second_head_tail[0].strip_edges()
			var second_tail = second_head_tail[1]
			# print("ï¼»åŠ©ç›£ï¼½ã€€second_headï¼š[" + second_head + "]")
			# print("ï¼»åŠ©ç›£ï¼½ã€€second_tailï¼š[" + second_tail + "]")
			# æ–‡å­—åˆ—ã®é…åˆ—ã«åˆ†å‰²
			var string_packed_array = second_head.split(":", true, 1)
			var instruction_code = string_packed_array[0] + ":"

			# ã‚³ãƒ¡ãƒ³ãƒˆ
			if second_head.begins_with("#"):
				pass

			else:
				# ä¾‹ãˆã° `img:` ã¨ã„ã£ãŸã‚³ãƒ¼ãƒ‰ã‹ã‚‰ã€ `ğŸ“—Img` ã¨ã„ã£ãŸå‘½ä»¤ãƒãƒ¼ãƒ‰ã‚’æ¤œç´¢ã—ã€ãã‚Œã‚’å®Ÿè¡Œã—ã¾ã™
				if instruction_code in self.directory_for_instruction_code_and_node_name:
					var instruction_node_name = self.directory_for_instruction_code_and_node_name[instruction_code]
					var instruction = self.hub().get_instruction(instruction_node_name)
					instruction.do_it(second_head)
				
			# ã•ã‚‰ã«å…ˆé ­è¡Œã‚’å–å¾—
			second_head_tail = self.hub().split_head_line_or_tail(second_tail)

		# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
		# ï¼»ãƒˆæ›¸ãï¼½çµ‚ã‚ã‚Š
		# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
		self.hub().is_department_not_found = false
		return true

	return false

